File: kaleido-gradle-plugin/src/main/java/com/tongsr/kaleido/gradle/RewriteSemanticXmlTask.java
package com.tongsr.kaleido.gradle;

import com.tongsr.kaleido.gradle.dsl.KaleidoProtectionDimension;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;
import javax.xml.XMLConstants;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import org.gradle.api.DefaultTask;
import org.gradle.api.file.ConfigurableFileCollection;
import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.file.RegularFileProperty;
import org.gradle.api.provider.Property;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.InputFile;
import org.gradle.api.tasks.InputFiles;
import org.gradle.api.tasks.OutputDirectory;
import org.gradle.api.tasks.OutputFile;
import org.gradle.api.tasks.PathSensitive;
import org.gradle.api.tasks.PathSensitivity;
import org.gradle.api.tasks.TaskAction;
import org.gradle.api.tasks.CacheableTask;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

@CacheableTask
public abstract class RewriteSemanticXmlTask extends DefaultTask {
    @Input public String getKaleidoCacheSchema() { return "SemanticXmlRewriteCache.v2"; }
    private static final String ANDROID = "http://schemas.android.com/apk/res/android";

    @Input public abstract Property<String> getConsumerProjectPath();
    @Input public abstract Property<String> getVariantName();
    @InputFile @PathSensitive(PathSensitivity.NONE)
    public abstract RegularFileProperty getAdoptionPlan();
    @InputFiles @PathSensitive(PathSensitivity.RELATIVE)
    public abstract ConfigurableFileCollection getConsumerResourceDirectories();
    @InputFiles @PathSensitive(PathSensitivity.RELATIVE)
    public abstract ConfigurableFileCollection getConsumerSourceDirectories();
    @OutputDirectory public abstract DirectoryProperty getOutputResources();
    @OutputFile public abstract RegularFileProperty getRewriteIntent();

    @TaskAction
    public void rewrite() throws Exception {
        var project = getConsumerProjectPath().get();
        var variant = getVariantName().get();
        var adoption = readProperties(Files.readAllBytes(
                getAdoptionPlan().get().getAsFile().toPath()));
        var applicationId = required(adoption, "applicationId", project, variant);
        var stream = required(adoption, "seed.domain.class-rewrite", project, variant);
        var protectedNames = new TreeSet<>(commaSeparated(
                adoption.get("protection.originalClassNames")));
        var hatches = RewriteClassesAndManifestTask.parseEscapeHatches(
                adoption.get("protection.classEscapeHatches"),
                EscapeHatchDeclaration.Kind.CLASS, project, variant);

        var documents = new TreeMap<String, XmlDocument>();
        var references = new ArrayList<XmlReference>();
        var seenPaths = new HashSet<String>();
        for (var root : getConsumerResourceDirectories().getFiles().stream()
                .map(java.io.File::toPath).filter(Files::isDirectory)
                .sorted(Comparator.comparing(Path::toString)).toList()) {
            try (var paths = Files.walk(root)) {
                for (var path : paths.filter(Files::isRegularFile)
                        .filter(file -> file.getFileName().toString().endsWith(".xml"))
                        .sorted().toList()) {
                    var relative = root.relativize(path).toString()
                            .replace(java.io.File.separatorChar, '/');
                    if (!seenPaths.add(relative)) {
                        throw RewriteClassesAndManifestTask.protectionFailure(
                                project, variant, relative,
                                "Semantic XML path is ambiguous across source roots",
                                "Resolve the source overlay before Kaleido rewriting");
                    }
                    var document = parse(Files.readAllBytes(path), project, variant, relative);
                    var found = inventory(document, relative, applicationId);
                    if (!found.isEmpty()) {
                        documents.put(relative, new XmlDocument(document, found));
                        references.addAll(found);
                    }
                }
            }
        }
        var roots = new TreeSet<String>();
        references.stream().map(XmlReference::resolvedIdentity)
                .filter(identity -> identity.startsWith(applicationId + "."))
                .forEach(roots::add);
        protectedNames.addAll(SourceProtectionScanner.inferProtectedIdentities(
                getConsumerSourceDirectories().getFiles(), roots));
        hatches.stream().filter(hatch ->
                        hatch.protects(KaleidoProtectionDimension.ORIGINAL_IDENTITY))
                .forEach(hatch -> roots.stream().filter(hatch::matches)
                        .forEach(protectedNames::add));
        var mapping = RewriteClassesAndManifestTask.allocateMapping(
                roots, protectedNames, roots, stream);
        var sites = references.stream()
                .filter(reference -> mapping.containsKey(reference.resolvedIdentity()))
                .map(reference -> new XmlSite(reference.location(), reference.lexicalValue(),
                        render(reference.lexicalValue(), applicationId,
                                mapping.get(reference.resolvedIdentity()))))
                .sorted(Comparator.comparing(XmlSite::location)).toList();
        var byLocation = sites.stream().collect(java.util.stream.Collectors.toMap(
                XmlSite::location, java.util.function.Function.identity()));
        var outputRoot = getOutputResources().get().getAsFile().toPath();
        deleteTree(outputRoot);
        for (var entry : documents.entrySet()) {
            var changed = false;
            for (var reference : entry.getValue().references()) {
                var site = byLocation.get(reference.location());
                if (site != null) {
                    reference.writer().set(site.target());
                    changed = true;
                }
            }
            if (changed) write(outputRoot.resolve(entry.getKey()), serialize(entry.getValue().document()));
        }
        var intent = new StringBuilder("schema=XmlRewriteIntent.v1\n");
        mapping.entrySet().stream().sorted(Map.Entry.comparingByKey()).forEach(entry ->
                intent.append("mapping=").append(entry.getKey()).append('|')
                        .append(entry.getValue()).append('\n'));
        sites.forEach(site -> intent.append("site=").append(site.location()).append('|')
                .append(site.original()).append('|').append(site.target()).append('\n'));
        write(getRewriteIntent().get().getAsFile().toPath(),
                intent.toString().getBytes(StandardCharsets.UTF_8));
    }

    private static List<XmlReference> inventory(
            Document document, String path, String applicationId) {
        var references = new ArrayList<XmlReference>();
        var elements = document.getElementsByTagName("*");
        for (var index = 0; index < elements.getLength(); index++) {
            var element = (Element) elements.item(index);
            var name = element.getTagName();
            if (name.contains(".") && !name.startsWith("android.")) {
                references.add(reference(path + "/element[" + index + "]", name,
                        value -> document.renameNode(element, element.getNamespaceURI(), value),
                        applicationId));
            }
            if (name.equals("view") || name.equals("fragment")) {
                addAttribute(references, path, index, element, null, "class", applicationId);
            }
            if (name.equals("fragment") || name.equals("activity")
                    || name.equals("dialog")) {
                addAttribute(references, path, index, element, ANDROID, "name", applicationId);
            }
            if (name.equals("variable") || name.equals("import")) {
                addAttribute(references, path, index, element, null, "type", applicationId);
            }
            var attributes = element.getAttributes();
            for (var attributeIndex = 0; attributeIndex < attributes.getLength(); attributeIndex++) {
                var attribute = attributes.item(attributeIndex);
                var local = attribute.getLocalName() == null
                        ? attribute.getNodeName() : attribute.getLocalName();
                if (local.equals("context") || local.equals("layout_behavior")
                        || local.equals("layoutManager") || local.equals("argType")) {
                    var value = attribute.getNodeValue();
                    if (looksLikeClass(value)) {
                        references.add(reference(path + "/element[" + index + "]@" + local,
                                value, attribute::setNodeValue, applicationId));
                    }
                }
            }
        }
        return references;
    }

    private static void addAttribute(
            List<XmlReference> references, String path, int index, Element element,
            String namespace, String attribute, String applicationId) {
        var present = namespace == null ? element.hasAttribute(attribute)
                : element.hasAttributeNS(namespace, attribute);
        if (!present) return;
        var value = namespace == null ? element.getAttribute(attribute)
                : element.getAttributeNS(namespace, attribute);
        if (!looksLikeClass(value)) return;
        references.add(reference(path + "/element[" + index + "]@" + attribute, value,
                replacement -> {
                    if (namespace == null) element.setAttribute(attribute, replacement);
                    else element.setAttributeNS(namespace, "android:" + attribute, replacement);
                }, applicationId));
    }

    private static XmlReference reference(
            String location, String lexical, StringWriter writer, String applicationId) {
        var resolved = lexical.startsWith(".") ? applicationId + lexical
                : lexical.contains(".") ? lexical : applicationId + "." + lexical;
        return new XmlReference(location, lexical, resolved, writer);
    }

    private static boolean looksLikeClass(String value) {
        return value.startsWith(".") || value.matches(
                "[A-Za-z_][A-Za-z0-9_]*(\\.[A-Za-z_$][A-Za-z0-9_$]*)+");
    }

    private static String render(String lexical, String applicationId, String target) {
        return lexical.startsWith(".") && target.startsWith(applicationId + ".")
                ? target.substring(applicationId.length()) : target;
    }

    private static Document parse(byte[] bytes, String project, String variant, String path) {
        try {
            var factory = DocumentBuilderFactory.newInstance();
            factory.setNamespaceAware(true);
            factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
            factory.setFeature("http://xml.org/sax/features/external-general-entities", false);
            factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
            factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
            factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
            return factory.newDocumentBuilder().parse(new java.io.ByteArrayInputStream(bytes));
        } catch (Exception failure) {
            throw RewriteClassesAndManifestTask.protectionFailure(project, variant, path,
                    "Consumer XML is not securely parseable",
                    "Fix or remove the malformed XML resource");
        }
    }

    private static byte[] serialize(Document document) throws Exception {
        var factory = TransformerFactory.newInstance();
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, "");
        var transformer = factory.newTransformer();
        transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");
        transformer.setOutputProperty(OutputKeys.INDENT, "no");
        var bytes = new java.io.ByteArrayOutputStream();
        transformer.transform(new DOMSource(document), new StreamResult(bytes));
        return bytes.toByteArray();
    }

    private static Map<String, String> readProperties(byte[] bytes) {
        var values = new HashMap<String, String>();
        for (var line : new String(bytes, StandardCharsets.UTF_8).split("\\n")) {
            var separator = line.indexOf('=');
            if (separator > 0) values.put(line.substring(0, separator), line.substring(separator + 1));
        }
        return Map.copyOf(values);
    }

    private static String required(
            Map<String, String> values, String key, String project, String variant) {
        var value = values.get(key);
        if (value == null) throw RewriteClassesAndManifestTask.protectionFailure(
                project, variant, key, "Adoption Plan is incomplete",
                "Regenerate a complete AdoptionPlan.v1");
        return value;
    }

    private static Set<String> commaSeparated(String value) {
        return value == null || value.isBlank() ? Set.of() : Set.of(value.split(","));
    }

    private static void deleteTree(Path root) throws IOException {
        if (!Files.exists(root)) return;
        try (var paths = Files.walk(root)) {
            for (var path : paths.sorted(Comparator.reverseOrder()).toList()) Files.delete(path);
        }
    }

    private static void write(Path path, byte[] bytes) throws IOException {
        Files.createDirectories(path.getParent());
        Files.write(path, bytes);
    }

    @FunctionalInterface private interface StringWriter { void set(String value); }
    private record XmlReference(String location, String lexicalValue,
                                String resolvedIdentity, StringWriter writer) {}
    private record XmlSite(String location, String original, String target) {}
    private record XmlDocument(Document document, List<XmlReference> references) {}
}

File: kaleido-gradle-plugin/src/main/kotlin/com/tongsr/kaleido/gradle/RewriteSemanticXmlTask.kt
package com.tongsr.kaleido.gradle

import com.tongsr.kaleido.gradle.dsl.KaleidoProtectionDimension
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.util.Comparator
import java.util.TreeMap
import java.util.TreeSet
import javax.xml.XMLConstants
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.OutputKeys
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult
import org.gradle.api.DefaultTask
import org.gradle.api.file.ConfigurableFileCollection
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.provider.Property
import org.gradle.api.tasks.CacheableTask
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.InputFile
import org.gradle.api.tasks.InputFiles
import org.gradle.api.tasks.OutputDirectory
import org.gradle.api.tasks.OutputFile
import org.gradle.api.tasks.PathSensitive
import org.gradle.api.tasks.PathSensitivity
import org.gradle.api.tasks.TaskAction
import org.w3c.dom.Document
import org.w3c.dom.Element

@CacheableTask
abstract class RewriteSemanticXmlTask : DefaultTask() {
    @get:Input
    val kaleidoCacheSchema: String
        get() = "SemanticXmlRewriteCache.v2"

    @get:Input
    abstract val consumerProjectPath: Property<String>

    @get:Input
    abstract val variantName: Property<String>

    @get:InputFile
    @get:PathSensitive(PathSensitivity.NONE)
    abstract val adoptionPlan: RegularFileProperty

    @get:InputFiles
    @get:PathSensitive(PathSensitivity.RELATIVE)
    abstract val consumerResourceDirectories: ConfigurableFileCollection

    @get:InputFiles
    @get:PathSensitive(PathSensitivity.RELATIVE)
    abstract val consumerSourceDirectories: ConfigurableFileCollection

    @get:OutputDirectory
    abstract val outputResources: DirectoryProperty

    @get:OutputFile
    abstract val rewriteIntent: RegularFileProperty

    @TaskAction
    fun rewrite() {
        val project = consumerProjectPath.get()
        val variant = variantName.get()
        val adoption = readProperties(Files.readAllBytes(adoptionPlan.get().asFile.toPath()))
        val applicationId = required(adoption, "applicationId", project, variant)
        val stream = required(adoption, "seed.domain.class-rewrite", project, variant)
        val protectedNames = TreeSet(commaSeparated(adoption["protection.originalClassNames"]))
        val hatches = RewriteClassesAndManifestTask.parseEscapeHatches(
            adoption["protection.classEscapeHatches"],
            EscapeHatchDeclaration.Kind.CLASS,
            project,
            variant,
        )

        val documents = TreeMap<String, XmlDocument>()
        val references = ArrayList<XmlReference>()
        val seenPaths = HashSet<String>()
        for (root in consumerResourceDirectories.files
            .map { it.toPath() }
            .filter { Files.isDirectory(it) }
            .sortedBy { it.toString() }) {
            Files.walk(root).use { paths ->
                for (path in paths.filter { Files.isRegularFile(it) }
                    .filter { file -> file.fileName.toString().endsWith(".xml") }
                    .sorted()
                    .toList()) {
                    val relative = root.relativize(path).toString()
                        .replace(File.separatorChar, '/')
                    if (!seenPaths.add(relative)) {
                        throw RewriteClassesAndManifestTask.protectionFailure(
                            project,
                            variant,
                            relative,
                            "Semantic XML path is ambiguous across source roots",
                            "Resolve the source overlay before Kaleido rewriting",
                        )
                    }
                    val document = parse(Files.readAllBytes(path), project, variant, relative)
                    val found = inventory(document, relative, applicationId)
                    if (found.isNotEmpty()) {
                        documents[relative] = XmlDocument(document, found)
                        references.addAll(found)
                    }
                }
            }
        }
        val roots = TreeSet<String>()
        references.map { it.resolvedIdentity }
            .filter { identity -> identity.startsWith("$applicationId.") }
            .forEach { roots.add(it) }
        protectedNames.addAll(
            SourceProtectionScanner.inferProtectedIdentities(
                consumerSourceDirectories.files,
                roots,
            ),
        )
        hatches.filter { hatch -> hatch.protects(KaleidoProtectionDimension.ORIGINAL_IDENTITY) }
            .forEach { hatch ->
                roots.filter { hatch.matches(it) }.forEach { protectedNames.add(it) }
            }
        val mapping = RewriteClassesAndManifestTask.allocateMapping(
            roots,
            protectedNames,
            roots,
            stream,
        )
        val sites = references
            .filter { reference -> mapping.containsKey(reference.resolvedIdentity) }
            .map { reference ->
                XmlSite(
                    reference.location,
                    reference.lexicalValue,
                    render(
                        reference.lexicalValue,
                        applicationId,
                        mapping.getValue(reference.resolvedIdentity),
                    ),
                )
            }
            .sortedBy { it.location }
        val byLocation = sites.associateBy { it.location }
        val outputRoot = outputResources.get().asFile.toPath()
        deleteTree(outputRoot)
        for (entry in documents.entries) {
            var changed = false
            for (reference in entry.value.references) {
                val site = byLocation[reference.location]
                if (site != null) {
                    reference.writer.set(site.target)
                    changed = true
                }
            }
            if (changed) {
                write(outputRoot.resolve(entry.key), serialize(entry.value.document))
            }
        }
        val intent = StringBuilder("schema=XmlRewriteIntent.v1\n")
        mapping.entries.sortedBy { it.key }.forEach { entry ->
            intent.append("mapping=").append(entry.key).append('|')
                .append(entry.value).append('\n')
        }
        sites.forEach { site ->
            intent.append("site=").append(site.location).append('|')
                .append(site.original).append('|').append(site.target).append('\n')
        }
        write(rewriteIntent.get().asFile.toPath(), intent.toString().toByteArray(StandardCharsets.UTF_8))
    }

    private fun interface StringWriter {
        fun set(value: String)
    }

    @JvmRecord
    private data class XmlReference(
        val location: String,
        val lexicalValue: String,
        val resolvedIdentity: String,
        val writer: StringWriter,
    )

    @JvmRecord
    private data class XmlSite(val location: String, val original: String, val target: String)

    @JvmRecord
    private data class XmlDocument(val document: Document, val references: List<XmlReference>)

    companion object {
        private const val ANDROID = "http://schemas.android.com/apk/res/android"

        private fun inventory(
            document: Document,
            path: String,
            applicationId: String,
        ): List<XmlReference> {
            val references = ArrayList<XmlReference>()
            val elements = document.getElementsByTagName("*")
            for (index in 0 until elements.length) {
                val element = elements.item(index) as Element
                val name = element.tagName
                if (name.contains(".") && !name.startsWith("android.")) {
                    references.add(
                        reference(
                            "$path/element[$index]",
                            name,
                            { value -> document.renameNode(element, element.namespaceURI, value) },
                            applicationId,
                        ),
                    )
                }
                if (name == "view" || name == "fragment") {
                    addAttribute(references, path, index, element, null, "class", applicationId)
                }
                if (name == "fragment" || name == "activity" || name == "dialog") {
                    addAttribute(references, path, index, element, ANDROID, "name", applicationId)
                }
                if (name == "variable" || name == "import") {
                    addAttribute(references, path, index, element, null, "type", applicationId)
                }
                val attributes = element.attributes
                for (attributeIndex in 0 until attributes.length) {
                    val attribute = attributes.item(attributeIndex)
                    val local = attribute.localName ?: attribute.nodeName
                    if (local == "context" || local == "layout_behavior" ||
                        local == "layoutManager" || local == "argType"
                    ) {
                        val value = attribute.nodeValue
                        if (looksLikeClass(value)) {
                            references.add(
                                reference(
                                    "$path/element[$index]@$local",
                                    value,
                                    { replacement -> attribute.nodeValue = replacement },
                                    applicationId,
                                ),
                            )
                        }
                    }
                }
            }
            return references
        }

        private fun addAttribute(
            references: MutableList<XmlReference>,
            path: String,
            index: Int,
            element: Element,
            namespace: String?,
            attribute: String,
            applicationId: String,
        ) {
            val present = if (namespace == null) {
                element.hasAttribute(attribute)
            } else {
                element.hasAttributeNS(namespace, attribute)
            }
            if (!present) return
            val value = if (namespace == null) {
                element.getAttribute(attribute)
            } else {
                element.getAttributeNS(namespace, attribute)
            }
            if (!looksLikeClass(value)) return
            references.add(
                reference(
                    "$path/element[$index]@$attribute",
                    value,
                    { replacement ->
                        if (namespace == null) {
                            element.setAttribute(attribute, replacement)
                        } else {
                            element.setAttributeNS(namespace, "android:$attribute", replacement)
                        }
                    },
                    applicationId,
                ),
            )
        }

        private fun reference(
            location: String,
            lexical: String,
            writer: StringWriter,
            applicationId: String,
        ): XmlReference {
            val resolved = when {
                lexical.startsWith(".") -> applicationId + lexical
                lexical.contains(".") -> lexical
                else -> "$applicationId.$lexical"
            }
            return XmlReference(location, lexical, resolved, writer)
        }

        private fun looksLikeClass(value: String): Boolean =
            value.startsWith(".") ||
                value.matches(Regex("[A-Za-z_][A-Za-z0-9_]*(\\.[A-Za-z_\$][A-Za-z0-9_\$]*)+"))

        private fun render(lexical: String, applicationId: String, target: String): String =
            if (lexical.startsWith(".") && target.startsWith("$applicationId.")) {
                target.substring(applicationId.length)
            } else {
                target
            }

        private fun parse(bytes: ByteArray, project: String, variant: String, path: String): Document {
            try {
                val factory = DocumentBuilderFactory.newInstance()
                factory.isNamespaceAware = true
                factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true)
                factory.setFeature("http://xml.org/sax/features/external-general-entities", false)
                factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false)
                factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "")
                factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "")
                return factory.newDocumentBuilder().parse(ByteArrayInputStream(bytes))
            } catch (failure: Exception) {
                throw RewriteClassesAndManifestTask.protectionFailure(
                    project,
                    variant,
                    path,
                    "Consumer XML is not securely parseable",
                    "Fix or remove the malformed XML resource",
                )
            }
        }

        private fun serialize(document: Document): ByteArray {
            val factory = TransformerFactory.newInstance()
            factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "")
            factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, "")
            val transformer = factory.newTransformer()
            transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8")
            transformer.setOutputProperty(OutputKeys.INDENT, "no")
            val bytes = ByteArrayOutputStream()
            transformer.transform(DOMSource(document), StreamResult(bytes))
            return bytes.toByteArray()
        }

        private fun readProperties(bytes: ByteArray): Map<String, String> {
            val values = HashMap<String, String>()
            for (line in String(bytes, StandardCharsets.UTF_8).split('\n')) {
                val separator = line.indexOf('=')
                if (separator > 0) {
                    values[line.substring(0, separator)] = line.substring(separator + 1)
                }
            }
            return values.toMap()
        }

        private fun required(
            values: Map<String, String>,
            key: String,
            project: String,
            variant: String,
        ): String {
            val value = values[key]
            if (value == null) {
                throw RewriteClassesAndManifestTask.protectionFailure(
                    project,
                    variant,
                    key,
                    "Adoption Plan is incomplete",
                    "Regenerate a complete AdoptionPlan.v1",
                )
            }
            return value
        }

        private fun commaSeparated(value: String?): Set<String> =
            if (value.isNullOrBlank()) emptySet() else value.split(',').toSet()

        private fun deleteTree(root: Path) {
            if (!Files.exists(root)) return
            Files.walk(root).use { paths ->
                for (path in paths.sorted(Comparator.reverseOrder()).toList()) Files.delete(path)
            }
        }

        private fun write(path: Path, bytes: ByteArray) {
            Files.createDirectories(path.parent)
            Files.write(path, bytes)
        }
    }
}
